src/config.rs

```diff
@@ -85,23 +106,82 @@ impl LayeredConfigs {
             .fold(config::Config::builder(), |builder, source| {
                 builder.add_source(source.clone())
             })
             .build()
             .expect("loaded configs should be merged without error")
     }
+
+    fn sources(&self) -> Vec<(ConfigSource, &config::Config)> {
+        let config_sources = [
+            (ConfigSource::Default, Some(&self.default)),
+            (ConfigSource::Env, Some(&self.env_base)),
+            (ConfigSource::User, self.user.as_ref()),
+            (ConfigSource::Repo, self.repo.as_ref()),
+            (ConfigSource::Env, Some(&self.env_overrides)),
+            (ConfigSource::CommandArg, self.arg_overrides.as_ref()),
+        ];
+        config_sources
+            .into_iter()
+            .filter_map(|(source, config)| config.map(|c| (source, c)))
+            .collect_vec()
+    }
+
+    pub fn resolved_config_values(
+        &self,
+        filter_prefix: &[&str],
+    ) -> Result<Vec<AnnotatedValue>, ConfigError> {
+        // Collect annotated values from each config.
+        let mut config_vals = vec![];
+
+        for (source, config) in self.sources() {
+            let config_table = config.collect()?.into();
+            let mut config_stack: Vec<(Vec<&str>, &config::Value)> = vec![(vec![], &config_table)];
+            while let Some((path, value)) = config_stack.pop() {
+                // Skip if path doesn't match filter_prefix.
+                if filter_prefix.iter().zip(path.iter()).any(|(a, b)| *a != *b) {
+                    continue;
+                }
+                match &value.kind {
+                    config::ValueKind::Table(table) => {
+                        // TODO: Remove sorting when config crate maintains deterministic ordering.
+                        for (k, v) in table.iter().sorted_by_key(|(k, _)| *k).rev() {
+                            let mut key_path = path.to_vec();
+                            key_path.push(k);
+                            config_stack.push((key_path, v));
+                        }
+                    }
+                    _ => {
+                        config_vals.push(AnnotatedValue {
+                            path: path.iter().map(|s| s.to_string()).collect_vec(),
+                            value: value.to_owned(),
+                            source: source.to_owned(),
+                            is_overridden: false,
+                        });
+                    }
+                }
+            }
+        }
+
+        // Walk through configs in order and mark each overridden value as overridden.
+        let mut keys_to_override = HashMap::new();
+        for val in &mut config_vals {
+            // Mark override on any previous value with same key.
+            if let Some(AnnotatedValue {
+                ref mut is_overridden,
+                ..
+            }) = keys_to_override.insert(val.path.to_owned(), val)
+            {
+                *is_overridden = true;
+            }
+        }
```

Comment:
I think I prefer the more concise (and easier to read IMO):

```rust
        for val in &mut config_vals {
            // Mark override on any previous value with same key.
            if let Some(mut value) = keys_to_override.insert(val.path.to_owned(), val) {
                value.is_overridden = true;
            }
        }
```

Reply:
And mark overridden from the top layer?

```rust
        let mut keys_found = HashSet::new();
        for val in config_vals.iter_mut().rev() {
            val.is_overridden = !keys_found.insert(&val.path);
        }
```

Reply:
Even better indeed.